module/VuFind/src/VuFind/Db/Service/TagService.php

```diff
@@ -778,4 +779,401 @@ public function getStatistics(bool $extended = false): array
+    ) {
+        $dql = 'SELECT DISTINCT(r.id) AS resource, r '
+            . 'FROM ' . $this->getEntityClass(Tags::class) . ' t '
+            . 'JOIN ' . $this->getEntityClass(ResourceTags::class) . ' rt WITH t.id = rt.tag '
+            . 'JOIN ' . $this->getEntityClass(Resource::class) . ' r WITH r.id = rt.resource '
+            . 'WHERE rt.resource IS NOT NULL ';
+        $parameters = [];
+        if ($fuzzy) {
+            $dql .= 'AND LOWER(t.tag) LIKE LOWER(:q) ';
+            $parameters['q'] = $q . '%';
+        } elseif (!$this->caseSensitive) {
+            $dql .= 'AND LOWER(t.tag) = LOWER(:q) ';
+            $parameters['q'] = $q;
+        } else {
+            $dql .= 'AND t.tag = :q ';
+            $parameters['q'] = $q;
+        }
+
+        if (!empty($source)) {
+            $dql .= 'AND r.source = :source';
+            $parameters['source'] = $source;
+        }
+
+        if (!empty($sort)) {
+            $dql = ResourceService::applySort($dql, $sort);
+        }
+
+        $query = $this->entityManager->createQuery($dql);
+        $query->setParameters($parameters);
+        if ($offset > 0) {
+            $query->setFirstResult($offset);
+        }
+        if (null !== $limit) {
+            $query->setMaxResults($limit);
+        }
+        $results = $query->getResult();
+        return $results;
+    }
+
+    /**
+     * Get tags associated with the specified resource.
+     *
+     * @param string $id          Record ID to look up
+     * @param string $source      Source of record to look up
+     * @param int    $limit       Max. number of tags to return (0 = no limit)
+     * @param int    $list        ID of list to load tags from (null for no
+     * restriction,  true for on ANY list, false for on NO list)
+     * @param int    $user        ID of user to load tags from (null for all users)
+     * @param string $sort        Sort type ('count' or 'tag')
+     * @param int    $userToCheck ID of user to check for ownership (this will
+     * not filter the result list, but rows owned by this user will have an is_me
+     * column set to 1)
+     *
+     * @return array
+     */
+    public function getForResource(
+        $id,
+        $source = DEFAULT_SEARCH_BACKEND,
+        $limit = 0,
+        $list = null,
+        $user = null,
+        $sort = 'count',
+        $userToCheck = null
+    ) {
+        $select = '';
+        $resourceTagsJoin = 'JOIN ';
+        $parameters = compact('id', 'source');
+        $tag = $this->caseSensitive ? 't.tag' : 'lower(t.tag)';
+        $fieldList = 't.id AS id, COUNT(DISTINCT(rt.user)) AS cnt, ' . $tag . ' AS tag';
+        // If we're looking for ownership, adjust query to include an "is_me" flag value indicating
+        // if the selected resource is tagged by the specified user.
+        if (!empty($userToCheck)) {
+            $fieldList .= ', MAX(CASE WHEN rt.user = :userToCheck THEN 1 ELSE 0 END) AS is_me';
+            $resourceTagsJoin = 'LEFT JOIN ';
```

Comment:
I'm not sure if there's a reason to do a LEFT JOIN here -- can't we just always do a regular JOIN? Since our WHERE clause always requires values from the Resource table, having matched values in the intermediate ResourceTags table seems necessary. The LEFT JOIN doesn't stop anything from working, but I'm pretty sure there's no need for it. But if I'm overlooking something, please let me know!

Reply:
You are right, as ResourceTags is the junction table either of the joins yields the same result.